AAInstrumentation/core.py

```diff
@@ -0,0 +1,135 @@
+from __future__ import annotations
+from subprocess import Popen, PIPE, run, DEVNULL
+import os
+import argparse
+from pathlib import Path, PosixPath
+from dataclasses import dataclass
+
+"""
+    This file contains the core functionality of the AAInstrumentation tool.
+    Specific kinds of instrumentation will inherit from AAInstrumentationDriver
+    and implement their specific functionality.
+"""
+
+
+@dataclass
+class AAInstrumentationDriver:
+
+    """
+    This class is the base class for all instrumentation drivers. It
+    contains the common functionality, like generating the baseline,
+    evaluating results, run a single instrumented compilation, etc. In
+    particular, this contains class variables for all the paths that have to
+    be specified to run the instrumentation on different machines and
+    benchmarks.
+    """
+
+    """
+    The path to the instrumented opt, clang, and llc binaries.
+    """
+    instr_path: Path
+
+    """
+    The directory where the instrumentation will be run.
+    """
+    exec_root: Path
+
+    """
+    The path to the specbuilder.
+    """
+    specbuild_dir: Path
+
+    """
+    The spec benchmark that is currently being instrumented.
+    """
+    benchmark: Path
+
+    """
+    The directory where the initial .bc files are located.
+    """
+    initial_dir: Path
+
+    """
+    The optimization level to be used on all compilations.
+    """
+    opt_flag: str
+
+    def generate_baseline(self):
+        run(
+            [
+                "./spec",
+                "build",
+                str(self.initial_dir),
+                str(self.benchmark),
+                str(self.instr_path.joinpath("clang")),
```

Comment:
does this work for c++ files?

Reply:
I never had in issue with this and I have tried to link before. Given that we will fundamentally change spec builder in the future, would it be fine to delay this issue until after that change?

Reply:
I can also open an issue or add a TODO note to the code